utils/src/types/unsigned_tx.rs

```diff
@@ -0,0 +1,521 @@
+                .as_slice()
+                .try_into()
+                .unwrap(),
+        );
+        UnlockScript::P2PKH(public_key)
+    }
+
+    fn input(hint: i32, key: &str, unlock_script: UnlockScript) -> TxInput {
+        let hint_bytes = i32::to_be_bytes(hint);
+        let hint = Hint::from_bytes(hint_bytes);
+        let key_bytes = hex_to_bytes(key).unwrap();
+        let key = Hash::from_bytes(key_bytes.as_slice().try_into().unwrap());
+        TxInput {
+            hint,
+            key,
+            unlock_script,
+        }
+    }
+
+    fn p2pkh_input(hint: i32, key: &str, public_key_hex: &str) -> TxInput {
+        let unlock_script = p2pkh_unlock_script(public_key_hex);
+        input(hint, key, unlock_script)
+    }
+
+    fn u256_from_str(amount_str: &str) -> U256 {
+        let number = BigUint::parse_bytes(amount_str.as_bytes(), 10).unwrap();
+        let mut bytes = number.to_bytes_be();
+        assert!(bytes.len() > 4);
+        let header: u8 = ((bytes.len() - 4) as u8) | 0xc0;
+        bytes.insert(0, header);
+        U256::from_encoded_bytes(&bytes)
+    }
+
+    fn output(amount_str: &str, lockup_script: LockupScript) -> AssetOutput {
+        let amount = u256_from_str(amount_str);
+        AssetOutput {
+            amount,
+            lockup_script,
+            lock_time: TimeStamp(0),
+            tokens: AVector::default(),
+            additional_data: ByteString::empty(),
+        }
+    }
+
+    fn token(token_id_hex: &str, amount_str: &str) -> Token {
+        let token_id_bytes = hex_to_bytes(token_id_hex).unwrap();
+        let token_id = Hash::from_bytes(token_id_bytes.as_slice().try_into().unwrap());
+        let amount = u256_from_str(amount_str);
+        Token {
+            id: token_id,
+            amount,
+        }
+    }
+
+    fn p2pkh_output_with_token(
+        amount_str: &str,
+        public_key_hash_hex: &str,
+        token_id_hex: &str,
+        token_amount: &str,
+    ) -> AssetOutput {
+        let output = p2pkh_output(amount_str, public_key_hash_hex);
+        let token = token(token_id_hex, token_amount);
+        AssetOutput {
+            tokens: AVector::from_item(token),
+            ..output
+        }
+    }
+
+    fn p2pkh_output(amount_str: &str, public_key_hash_hex: &str) -> AssetOutput {
+        output(amount_str, p2pkh_lockup_script(public_key_hash_hex))
+    }
+
+    fn decode_and_check_tx(
+        tx_id_hex: &str,
+        encoded_tx: Vec<u8>,
+        network_id: u8,
+        gas_amount: I32,
+        gas_price: U256,
+        is_script_tx: bool,
+        all_inputs: &[TxInput],
+        all_outputs: &[AssetOutput],
+    ) {
+        let check = |tx: &UnsignedTx| match tx {
+            UnsignedTx::Version(byte) => assert_eq!(byte.0, 0),
+            UnsignedTx::NetworkId(byte) => assert_eq!(byte.0, network_id),
+            UnsignedTx::ScriptOpt(script) => {
+                if is_script_tx {
+                    assert!(script.inner.is_some())
```

Comment:
[Nit] Would be great to have a test that test the instrs explicitly as well.

Reply:
Due to the limitations of stack memory size, we don't save all instrs. In `AVector`, we only save the item currently being decoded. Once decoding is completed, we display it(if necessary) and then discard the item.

Reply:
I see, thanks for the context, that makes sense :+1: